### ISSUE TITLE
Find current lexer in formatter

### ISSUE THREAD

**Issue by asbjornu** [rouge-ruby/rouge#1463]
I have defined a lexer for PlantUML diagrams as such:

```ruby
module Rouge
  module Lexers
    class PlantUml < PlainText
      title "A passthrough lexer used for PlantUML input"
      tag 'plantuml'
    end
  end
end
```

I've also subclassed `Rouge::Formatters::HTML` so I can replace the plain text diagram with rendered SVG when a `plantuml` codefence is found. I'm stuck, though, since I can't figure out which lexer the `tokens` I'm passed have been lexed by:
```ruby
module Rouge
  module Formatters
    class PlantUml < HTML
      tag 'platnuml'

      def stream(tokens, &block)
        p tokens.inspect
        yield %(<div class="highlight"><pre class="highlight"><code>)
        super
        yield "</code></pre></div>"
      end
  end
end
```

The result from `p tokens.inspect` is:

```ruby
"#<Enumerator: #<Rouge::Formatters::PlantUmlFormatter:0x00007f8c04551198>:filter_escapes(#<Enumerator: #<Rouge::Lexers::PlantUml:0x00007f8c045512d8 @options={}, @debug=false, @token=<Token Text>>:lex(\"@startuml\\nactor client\\nnode app\\ndatabase db\\n\\ndb -> app\\napp -> client\\n@enduml\\n\")>)>"
```

So I can see from the inspection that the current lexer is `Rouge::Lexers::PlantUml`, but I can't for the life of me figure out how to extract that information from `tokens` without doing simple string search:

```ruby
if tokens.inspect.include? 'Rouge::Lexers::PlantUml'
  # We've found a PlantUML diagram through a dirty string search
end
```

Is there another way?

I'm also having difficulties getting the default formatting behavior of Rouge when I subclass `HTML` as per above. What I'd like is for my custom PlantUML formatter to act like it didn't exist except when it finds a PlantUML diagram.

If I'm going about this the wrong way, please let me know. I'm quite new to both Ruby and Rouge, so I'm sorry for being so confused.

**Comment by jneen** [comment]
Hi! You should not use Rouge for rendering PlantUml. Instead, add a condition to your Markdown processor.

**Comment by jneen** [comment]
In simpler terms, when a PlantUML section is encountered, you should take steps to prevent it from even getting to a Rouge layer. This is the approach GitLab takes with both PlantUML and embedded LaTeX.

**Comment by jneen** [comment]
If you are using the built-in Redcarpet plugin, you can try:

``` ruby
module MyRedcarpetPlugin
  include Rouge::Redcarpet
  def block_code(code, language)
    case language.downcase
    when 'plantuml', 'uml' #, etc...
      somehow_render_plantuml_to_html(code)
    when 'math'
      somehow_render_latex_to_html(code)
    else
      super # call into rouge, render highlighted content
    end
  end
end
```
  end
```

**Comment by asbjornu** [comment]
Thank you so much for the lightning speed of your reply and the accuracy and wealth of information you provide, @jneen! ⚡️  Much, much appreciated! 🙏 

I don't think I'm currently using Redcarpet, unless that's included by default in Jekyll somehow. I think I'm just using the default `markdown: kramdown` configuration. I'm also unsuccessful getting your code to execute. I called my module `Redplant` and this is the first error I had:

```ruby
uninitialized constant Redplant::Rouge
```

Adding `require 'rouge'` at the top gave me this:

```ruby
uninitialized constant Rouge::Redcarpet
```

Then adding `require 'redcarpet'` gives me the following error even after doing `gem install redcarpet`:

```ruby
 Dependency Error: Yikes! It looks like you don't have _plugins/redplant.rb or one of its dependencies installed. In order to use Jekyll as currently configured, you'll need to install this gem. The full error message from Ruby is: 'cannot load such file -- redcarpet' If you run into trouble, you can find helpful resources at https://jekyllrb.com/help/!
```

Ideas?

**Comment by jneen** [comment]
I would ask for support over on the Jekyll side. They almost 100% will have some kind of plugin that you can inject code into.

**Comment by jneen** [comment]
(kramdown and redcarpet are competing implementations of markdown, so you'll have to do this in kramdown instead)

**Comment by jneen** [comment]
Okay, looking at it more closely, it appears that kramdown is optimized for speed at the cost of flexibility - you might want to switch to redcarpet for this. If you *really* want to stick to kramdown you may have to take a monkey-patching approach like in this (unmaintained) project: https://github.com/tmthrgd/jekyll-graphs/blob/master/jekyll-graphs.rb#L547

**Comment by asbjornu** [comment]
I see. Thank you so much for your pointers, knowledge and time, @jneen! :pray: I have no idea what I lose or gain by changing from kramdown to redcarpet, but I wouldn't mind changing if I get this to work. The problem is that I have the dependency error even after configuring `markdown: redcarpet` or `markdown: redplant`.

**Comment by jneen** [comment]
As for your errors, I'm not sure I would be able to say much without seeing your code. I am not a user or maintainer of Jekyll, and I do know they do a significant amount of black magic that can cause surprising behaviour.

**Comment by asbjornu** [comment]
This is how my class looks like right now:

```ruby
require 'redcarpet'
require 'rouge'
require 'rouge/plugins/redcarpet'

module Redplant
  include Rouge::Redcarpet

  def block_code(code, language)
    case language.downcase
      when 'plantuml'
        puts 'We have PlantUML'
      else
        super # call into rouge, render highlighted content
    end
  end
end
```

This code resides in a `redplant.rb` file within Jekyll's `_plugins` folder. Your guess is as good as mine, but if you insist this is a Jekyll problem, I'll move over there with my questions.

**Comment by asbjornu** [comment]
Thanks to your tip about monkeypatching Kramdown, I've now gotten the following to work:

```ruby
require 'shell'

class Kramdown::Converter::Html
  alias_method :super_convert_codeblock, :convert_codeblock

  def convert_codeblock el, indent
    if el.attr["class"] != "language-plantuml"
      return super_convert_codeblock(el, indent)
    end

    return convert_plantuml_to_svg(el.value)
  end

  def convert_plantuml_to_svg(content)
    plant_uml_jar_file = "./_plugins/plantuml.1.2020.2.jar"

    sh = Shell.new
    svg = (
      sh.echo(content) |
      # We need to capture the std and err output from the Java process
      # because even if conversion fails due to missing dot executable
      # or other problems, the exit code seems to be 0.
      sh.system("java -jar #{plant_uml_jar_file} -tsvg -pipe")
    ).to_s

    return svg
  end
end
```

Thank you so much! 🙏 ❤️
